File: apps/web-lite/src/analysis/classify.ts
/**
 * ルールベース分類。docs 01 §5 の TypeScript 例 + docs 02 の
 * wavetable candidate 判定(conf>0.7, stability<20cents, dur>300ms)。
 */
import type { AssetType } from "../manifest/schema";
import type { SliceFeatures } from "./features";

const SILENCE_PEAK_DB = -55;

export function classify(features: SliceFeatures): AssetType {
  if (features.peakDb < SILENCE_PEAK_DB) {
    return "Reject";
  }

  if (
    features.durationSec < 1.2 &&
    features.attackMs < 35 &&
    (features.pitchConfidence ?? 0) < 0.45
  ) {
    return "PercussiveOneShot";
  }

  // wavetable candidate: pitch が十分安定して持続する音
  if (
    features.durationSec > 0.3 &&
    (features.pitchConfidence ?? 0) > 0.7 &&
    (features.pitchStabilityCents ?? 999) < 20 &&
    (features.voicedRatio ?? 0) > 0.6
  ) {
    return "WavetableCandidate";
  }

  if (
    features.durationSec < 2.5 &&
    (features.pitchConfidence ?? 0) > 0.65 &&
    (features.pitchStabilityCents ?? 999) < 25
  ) {
    return "MelodicOneShot";
  }

  if (
    features.durationSec > 3.0 &&
    features.transientDensity < 1.0 &&
    (features.pitchConfidence ?? 0) > 0.6
  ) {
    return "DroneLoop";
  }

  if (
    features.durationSec > 3.0 &&
    features.spectralFlatness > 0.45 &&
    features.transientDensity < 1.0
  ) {
    return "NoiseTexture";
  }

  return "MelodicPhrase";
}

/** 分類の確からしさの簡易スコア。ルールベースなので目安。 */
export function classificationConfidence(
  type: AssetType,
  features: SliceFeatures
): number {
  switch (type) {
    case "Reject":
      return 0.9;
    case "PercussiveOneShot": {
      const fast = Math.max(0, 1 - features.attackMs / 35);
      const unpitched = 1 - (features.pitchConfidence ?? 0);
      return Math.min(0.95, 0.5 + 0.3 * fast + 0.2 * unpitched);
    }
    case "WavetableCandidate":
      return Math.min(0.95, features.pitchConfidence ?? 0.5);
    case "MelodicOneShot":
      return Math.min(0.9, features.pitchConfidence ?? 0.5);
    case "DroneLoop":
      return Math.min(0.85, 0.4 + (features.pitchConfidence ?? 0) * 0.5);
    case "NoiseTexture":
      return Math.min(0.85, 0.4 + features.spectralFlatness * 0.8);
    default:
      return 0.5;
  }
}

File: apps/backend/app/pipeline/classify.py
"""分類: ルールベース大分類 + stem ヒント。docs 03 §6 / docs 07 §6 の方針
(AI は決定ではなく候補提示 — CLAP タグは features["clapTags"] に付くだけ)。
"""
from __future__ import annotations

SILENCE_PEAK_DB = -55.0


def classify(features: dict, stem: str | None = None) -> str:
    dur = features.get("durationSec", 0.0)
    peak = features.get("peakDb", -100.0)
    attack = features.get("attackMs", 999.0)
    conf = features.get("f0Confidence") or 0.0
    stab = features.get("f0StabilityCents")
    stab = 999.0 if stab is None else stab
    voiced = features.get("voicedRatio") or 0.0
    flat = features.get("spectralFlatnessMean") or 0.0
    td = features.get("transientDensity") or 0.0

    if peak < SILENCE_PEAK_DB:
        return "Reject"

    # stem ヒント (docs 01 §2: drums→percussive, bass→bass, vocals→vocal chop)
    # 注: pyin の confidence (voiced_prob 平均) は実楽曲で 0.3〜0.45 程度に
    # 留まるため、Web Lite (YIN) より閾値を下げてある。
    if stem == "drums" and dur < 1.5:
        return "PercussiveOneShot"
    if stem == "vocals":
        return "VocalChop" if dur < 2.0 else "MelodicPhrase"
    if stem == "bass" and conf > 0.25 and voiced > 0.6 and dur < 2.5:
        return "BassOneShot"

    if dur < 1.2 and attack < 35 and conf < 0.25:
        return "PercussiveOneShot"

    if dur > 0.15 and conf > 0.35 and stab < 20 and voiced > 0.6:
        return "WavetableCandidate"

    if dur < 2.5 and conf > 0.3 and stab < 25 and voiced > 0.5:
        return "MelodicOneShot"

    if dur > 3.0 and td < 1.0 and conf > 0.3:
        return "DroneLoop"

    if dur > 3.0 and flat > 0.45 and td < 1.0:
        return "NoiseTexture"

    if dur > 3.0 and td < 1.5 and voiced < 0.3:
        return "AmbienceLoop"

    return "MelodicPhrase"


def classification_confidence(asset_type: str, features: dict) -> float:
    conf = features.get("f0Confidence") or 0.0
    flat = features.get("spectralFlatnessMean") or 0.0
    attack = features.get("attackMs", 999.0)

    if asset_type == "Reject":
        return 0.9
    if asset_type == "PercussiveOneShot":
        fast = max(0.0, 1 - attack / 35)
        return min(0.95, 0.5 + 0.3 * fast + 0.2 * (1 - conf))
    if asset_type == "WavetableCandidate":
        return min(0.95, conf)
    if asset_type in ("MelodicOneShot", "BassOneShot"):
        return min(0.9, max(0.5, conf))
    if asset_type == "DroneLoop":
        return min(0.85, 0.4 + conf * 0.5)
    if asset_type in ("NoiseTexture", "AmbienceLoop"):
        return min(0.85, 0.4 + flat * 0.8)
    return 0.5
